sql/database.go

```diff
@@ -226,9 +230,15 @@ func (db *Database) Exec(query string, encoder Encoder, decoder Decoder) (int, e
 
 // Close closes all pooled connections.
 func (db *Database) Close() error {
+	db.closeMux.Lock()
+	defer db.closeMux.Unlock()
+	if db.closed {
+		return nil
+	}
 	if err := db.pool.Close(); err != nil {
 		return fmt.Errorf("close pool %w", err)
 	}
+	db.closed = true
 	return nil
 }
```

Comment:
I thought about that to, but that doesn't work if `db.pool.Close` returns an error. In that case close cannot call `db.pool.Close` again:

```golang
// Close closes all pooled connections.
func (db *Database) Close() error {
	var err error
	db.closeOncer.Do(func() {
		err = db.pool.Close()
	})
	return err
}
```

Reply:
You could persist the close error on the struct, then that approach would be a little neater than the current approach.

Reply:
Yes in the case of returning the error that would work, but if `db.pool.Close()` was not successful another call to `Database.Close()` should try again in my opinion instead of returning the error from the last try.